src/main/java/bio/terra/service/snapshotbuilder/SnapshotRequestDao.java

```diff
@@ -0,0 +1,186 @@
+      throw new NotFoundException("No dataset found for given id", ex);
+    }
+  }
+
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public SnapshotAccessRequestResponse create(
+      UUID datasetId, SnapshotAccessRequest request, String email) {
+    String jsonValue;
+    try {
+      jsonValue = objectMapper.writeValueAsString(request.getDatasetRequest());
+    } catch (JsonProcessingException e) {
+      throw new BadRequestException("Could not write snapshot access request to json", e);
+    }
+    DaoKeyHolder keyHolder = new DaoKeyHolder();
+    String sql =
+        """
+        INSERT INTO snapshot_requests
+        (dataset_id, request_name, request_research_purpose, snapshot_builder_request, user_email)
+        VALUES (:dataset_id, :request_name, :request_research_purpose, cast(:snapshot_builder_request as jsonb), :user_email)
+        """;
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue(datasetIdField, datasetId)
+            .addValue(requestNameField, request.getName())
+            .addValue(requestResearchPurposeField, request.getResearchPurposeStatement())
+            .addValue(snapshotBuilderRequestField, jsonValue)
+            .addValue(userEmailField, email);
+    try {
+      jdbcTemplate.update(sql, params, keyHolder);
+    } catch (DataIntegrityViolationException ex) {
+      throw new NotFoundException("Dataset with given dataset id does not exist.");
+    }
+    UUID id = keyHolder.getId();
+    return getById(id);
+  }
+
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public SnapshotAccessRequestResponse update(
+      UUID requestId, SnapshotAccessRequestResponse.StatusEnum status) {
+    String sql =
+        """
+        UPDATE snapshot_requests SET
+        status = :status, update_date = :update_date
+        WHERE id = :id
+        """;
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue(statusField, status.toString())
+            .addValue(updateDateField, Timestamp.from(Instant.now()))
+            .addValue(requestIdField, requestId);
+    if (jdbcTemplate.update(sql, params) == 0) {
+      throw new NotFoundException("Snapshot Access Request with given id does not exist.");
+    }
+    return getById(requestId);
+  }
+
+  @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+  public void delete(UUID requestId) {
+    String sql = "DELETE FROM snapshot_requests WHERE id = :id";
+    MapSqlParameterSource params = new MapSqlParameterSource().addValue(requestIdField, requestId);
+    try {
+      if (jdbcTemplate.update(sql, params) == 0) {
+        throw new NotFoundException("Snapshot Request with given id does not exist.");
+      }
+    } catch (EmptyResultDataAccessException ex) {
+      throw new NotFoundException("No snapshot request found for given id", ex);
+    }
```

Comment:
These two errors are pretty close to one another.  Is that intentional / necessary?